docs/hub/datasets-spark.md

```diff
@@ -0,0 +1,441 @@
+
+```python
+>>> sel = sel = [("langdetect", "=", "zh-cn")]
+>>> df_chinese_only = read_parquet("hf://datasets/BAAI/Infinity-Instruct/7M/*.parquet", filters=sel)
+>>> df_chinese_only
++---+----------------------------+-----+----------+----------+                  
+| id|               conversations|label|langdetect|    source|
++---+----------------------------+-----+----------+----------+
+|  9|[{human, 因果联系原则是法...|     |     zh-cn|Subjective|
+| 19|[{human, 用经济学知识分析...|     |     zh-cn|Subjective|
+| 38| [{human, 某个考试共有A、...|     |     zh-cn|Subjective|
+| 39|[{human, 撰写一篇关于斐波...|     |     zh-cn|Subjective|
+| 57|[{human, 总结世界历史上的...|     |     zh-cn|Subjective|
+| 61|[{human, 生成一则广告词。...|     |     zh-cn|Subjective|
+| 66|[{human, 描述一个有效的团...|     |     zh-cn|Subjective|
+| 94|[{human, 如果比利和蒂芙尼...|     |     zh-cn|Subjective|
+|102|[{human, 生成一句英文名言...|     |     zh-cn|Subjective|
+|106|[{human, 写一封感谢信，感...|     |     zh-cn|Subjective|
+|118| [{human, 生成一个故事。}...|     |     zh-cn|Subjective|
+|174|[{human, 高胆固醇水平的后...|     |     zh-cn|Subjective|
+|180|[{human, 基于以下角色信息...|     |     zh-cn|Subjective|
+|192|[{human, 请写一篇文章，概...|     |     zh-cn|Subjective|
+|221|[{human, 以诗歌形式表达对...|     |     zh-cn|Subjective|
+|228|[{human, 根据给定的指令，...|     |     zh-cn|Subjective|
+|236|[{human, 打开一个新的生成...|     |     zh-cn|Subjective|
+|260|[{human, 生成一个有关未来...|     |     zh-cn|Subjective|
+|268|[{human, 如果有一定数量的...|     |     zh-cn|Subjective|
+|273| [{human, 题目：小明有5个...|     |     zh-cn|Subjective|
++---+----------------------------+-----+----------+----------+
+```
+
+## Write
+
+We also provide a helper function to write datasets in a distributed manner to your Hugging Face account.
+
+You can write a PySpark Dataframe to Hugging Face using this `write_parquet` helper functions based on the `huggingface_hub` API.
+In particular it uses the `preupload_lfs_files` utility to upload Parquet files in parallel in a distributed manner, and only commit the files once they're all uploaded:
+
+
+```python
+import math
+import pickle
+import tempfile
+from functools import partial
+from typing import Iterator, Optional
+from uuid import uuid4
+
+import pyarrow as pa
+import pyarrow.parquet as pq
+from huggingface_hub import CommitOperationAdd, CommitOperationCopy, CommitOperationDelete, HfFileSystem
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.pandas.types import from_arrow_schema, to_arrow_schema
+
+
+def _write_single(iterator: Iterator[pa.RecordBatch], path: str, schema: pa.Schema, filesystem: HfFileSystem, row_group_size: Optional[int] = None, **kwargs) -> Iterator[pa.RecordBatch]:
+    resolved_path = filesystem.resolve_path(path)
+    with tempfile.NamedTemporaryFile(suffix=".parquet") as temp_file:
+        with pq.ParquetWriter(temp_file.name, schema=schema, **kwargs) as writer:
+            for batch in iterator:
+                writer.write_batch(batch, row_group_size=row_group_size)
+        filesystem._api.upload_file(
```

Comment:
Can't we use a `filesystem.open("w").write(...).flush()` or similar? Feels weird to introduce `filesystem` but to rely on `HfApi` in the end.

Same for `_rename` / `_preupload` / `_commit` where we can't really do without `HfApi`. What feels weird to me is that we use `filesystem` only for `filesystem.resolve_path(path)`. In the end, we could also just define a method

```py
def write_parquet(df: DataFrame, repo_id: str, path_in_repo: str, *, repo_type: Optional[str] = None, revision: Optional[str] = None, **kwargs) -> None:
```

Reply:
The API choice is mostly to fit the pyspark API which uses urls like s3://

Then it would be cool to use the filesystem only, but I don't think the preupload_lfs logic is available (it's the only way to do parallel uploads afaik)